### PR TITLE
build: add bazel build files to most packages

### PR DESCRIPTION
**Pull request by alexeagle**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,15 +24,16 @@
   "dependencies": {
     "core-js": "^2.4.1",
     "reflect-metadata": "^0.1.3",
-    "rxjs": "^5.5.2",
+    "rxjs": "git+https://github.com/gregmagolan/rxjs.git#rxjs-bazel-dist-2",
```
Comment (IgorMinar) [inline package.json]: Are these changes being upstreamed? Can you link to the PR from the commit message?

File: .circleci/config.yml [hunk 2]
```diff
@@ -42,7 +42,7 @@ jobs:
 
       - run: bazel run @yarn//:yarn
       - run: bazel build packages/...
-      - run: bazel test @angular//...
+      - run: bazel test packages/... @angular/...
```
Comment (IgorMinar) [inline .circleci/config.yml]: what is `@angular/...` that is not under packages/... ?

File: packages/BUILD.bazel [hunk 3]
```diff
@@ -1 +1,20 @@
...
+    "tsconfig-build.json",
+    "tsconfig.json",
+])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_config", "ts_library")
+
+ts_library(
+    name = "types",
+    srcs = glob(["*.ts"]),
+    tsconfig = "tsconfig.json",
```
Comment (IgorMinar) [inline packages/BUILD.bazel]: so unlike in g3, externally we allow tsconfig.json to be specified?

Comment (IgorMinar) [reply packages/BUILD.bazel]: I'm not opposed to it. I think that's the only way externally, I'm just wondering if this is intentional.

File: packages/animations/browser/testing/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/animations/browser/testing/BUILD.bazel]: "/testing" secondary entry point is not `testonly` - it's prod code that we ship to npm.

tests are only those files that end with `_spec.ts` or `.spec.ts` typically in the `test/` directory. eventually, I'd prefer to mix the src and test code so that we don't have two parallel file trees but that's not what you should be trying to do here.

File: packages/bazel/test/ngc-wrapped/test_support.ts [hunk 5]
```diff
@@ -150,7 +150,7 @@ function makeTempDir(baseDir): string {
 }
 
 export function listFilesRecursive(dir: string, fileList: string[] = []) {
-  fs.readdirSync(dir).map(file => {
+  fs.readdirSync(dir).forEach(file => {
```
Comment (IgorMinar) [inline packages/bazel/test/ngc-wrapped/test_support.ts]: is this less surprising (since map returns something that is not used) or why this change?

File: packages/common/BUILD.bazel [hunk 6]
```diff
@@ -1,15 +1,19 @@
-package(default_visibility=["//visibility:public"])
+package(default_visibility = ["//visibility:public"])
```
Comment (IgorMinar) [inline packages/common/BUILD.bazel]: finished reviewing up to here. getting of the bus, so I'm posting what I have for now.

**Review by IgorMinar — COMMENTED** [review]
posting review part 1

File: packages/common/http/testing/BUILD.bazel [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/common/http/testing/BUILD.bazel]: nope

File: packages/common/testing/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/common/testing/BUILD.bazel]: nope

File: packages/compiler/test/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "test",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/compiler/test/BUILD.bazel]: yup :)

File: packages/compiler/test/BUILD.bazel [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        "//packages:types",
+        "//packages/common",
+        "//packages/compiler",
+        "//packages/compiler-cli",
+        "//packages/compiler/testing",
+        "//packages/core",
+        "//packages/core/testing",
+        "//packages/platform-browser",
+        "//packages/platform-browser-dynamic",
+        "//packages/platform-browser/testing",
```
Comment (IgorMinar) [inline packages/compiler/test/BUILD.bazel]: how did you determine this dep list? trial and error? something else?

File: packages/compiler/testing/BUILD.bazel [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/compiler/testing/BUILD.bazel]: nope

File: packages/core/testing/BUILD.bazel [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/core/testing/BUILD.bazel]: nope

File: packages/core/testing/BUILD.bazel [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    srcs = glob(["**/*.ts"]),
+    module_name = "@angular/core/testing",
+    tsconfig = "//packages:tsconfig",
+    deps = [
+        "//packages:types",
```
Comment (IgorMinar) [inline packages/core/testing/BUILD.bazel]: what's this one?

File: packages/http/testing/BUILD.bazel [hunk 14]
```diff
@@ -0,0 +1,16 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/http/testing/BUILD.bazel]: nope

File: packages/common/BUILD.bazel [hunk 15]
```diff
@@ -1,15 +1,19 @@
...
+
 load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
 
 ts_library(
```
Comment (IgorMinar) [inline packages/common/BUILD.bazel]: shouldn't this be an ng_module rule?

Comment (IgorMinar) [reply packages/common/BUILD.bazel]: also elsewhere...

File: packages/platform-browser-dynamic/BUILD.bazel [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+ts_library(
+    name = "platform-browser-dynamic",
+    srcs = glob(
+        [
+            "*.ts",
+            "src/**/*.ts",
+        ],
+    ),
+    module_name = "@angular/platform-browser-dynamic",
+    tsconfig = "//packages:tsconfig",
```
Comment (IgorMinar) [inline packages/platform-browser-dynamic/BUILD.bazel]: is it right that you are reusing this single tsconfig file for all the ts_libraries? we have package specific tsconfigs already, so I'm surprised that you are not using those..

File: packages/platform-browser-dynamic/testing/BUILD.bazel [hunk 17]
```diff
@@ -0,0 +1,20 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/platform-browser-dynamic/testing/BUILD.bazel]: nope

File: packages/platform-browser/testing/BUILD.bazel [hunk 18]
```diff
@@ -0,0 +1,16 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/platform-browser/testing/BUILD.bazel]: nope

File: packages/platform-server/testing/BUILD.bazel [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+ts_library(
+    name = "testing",
+    testonly = 1,
+    srcs = glob(["**/*.ts"]),
+    module_name = "@angular/platform-server/testing",
+    tsconfig = "//packages:tsconfig",
+    deps = [
+        "//packages/core",
+        "//packages/platform-browser-dynamic/testing",
+        "//packages/platform-browser/animations",
```
Comment (IgorMinar) [inline packages/platform-server/testing/BUILD.bazel]: animations? really?

File: packages/router/testing/BUILD.bazel [hunk 20]
```diff
@@ -0,0 +1,17 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/router/testing/BUILD.bazel]: nope

File: packages/service-worker/BUILD.bazel [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+    srcs = glob(
+        [
+            "*.ts",
+            "src/**/*.ts",
+        ],
+    ),
+    module_name = "@angular/service-worker",
+    tsconfig = "//packages:tsconfig",
+    deps = [
+        "//packages/core",
```
Comment (IgorMinar) [inline packages/service-worker/BUILD.bazel]: I think we now also depend on common. we just merged a PR that does that.

File: packages/service-worker/worker/testing/BUILD.bazel [hunk 22]
```diff
@@ -0,0 +1,14 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
Comment (IgorMinar) [inline packages/service-worker/worker/testing/BUILD.bazel]: nope

File: packages/tsconfig-build.json [hunk 23]
```diff
@@ -23,5 +23,9 @@
     "emitDecoratorMetadata": true,
     "sourceMap": true,
     "inlineSources": true
+  },
+
+  "bazelOptions": {
+    "disableStrictDeps": true
```
Comment (IgorMinar) [inline packages/tsconfig-build.json]: can you leave a comment in here to explain why this is turned off?

File: yarn.lock [hunk 24]
```diff
@@ -7156,6 +7161,12 @@ tsscmp@1.0.5:
   version "1.0.5"
   resolved "https://registry.yarnpkg.com/tsscmp/-/tsscmp-1.0.5.tgz#7dc4a33af71581ab4337da91d85ca5427ebd9a97"
 
+tsutils@2.12.1:
```
Comment (IgorMinar) [inline yarn.lock]: what is this one here for? I didn't notice it being used directly by any of your changes....

**Review by IgorMinar — CHANGES_REQUESTED** [review]
done with the first full review pass..

File: .circleci/config.yml (hunk 2 above)
Comment (alexeagle) [reply .circleci/config.yml]: @angular is the name of the workspace that lives in packages/bazel. That's how users can
`load("@angular//:index.bzl", "ng_module")`

The @angular_src workspace lives in the root dir.

The outer workspace doesn't see anything inside the nested workspace.

File: package.json (hunk 1 above)
Comment (alexeagle) [reply package.json]: yes, https://github.com/ReactiveX/rxjs/pull/3131

File: packages/BUILD.bazel (hunk 3 above)
Comment (alexeagle) [reply packages/BUILD.bazel]: correct.
Before landing this PR I'd like to fix https://github.com/bazelbuild/rules_typescript/issues/93 so we don't have to specify this everywhere.
Note that the BUILD files only ever refer to one tsconfig file, so after fully switching to Bazel we can delete the others. The information they contain is just the dependency graph.

File: packages/animations/browser/testing/BUILD.bazel (hunk 4 above)
Comment (alexeagle) [reply packages/animations/browser/testing/BUILD.bazel]: That's not what `testonly` means. It's applied both to tests, and to test-support code.

`testonly` is a marker that prevents a non-testonly target from depending on a `testonly` one - it taints the build graph and bazel enforces this.

We can still ship `testonly` packages to NPM. what this does is prevent mistakes like `//packages/common` depending on `//packages/core/testing`, and also when Angular/Bazel users make a source dependency on Angular, they'll get the same constraint so they don't ship testing code.

File: packages/bazel/test/ngc-wrapped/test_support.ts (hunk 5 above)
Comment (alexeagle) [reply packages/bazel/test/ngc-wrapped/test_support.ts]: this change is required.
This doesn't type-check under Bazel, because it violates http://tsetse.info/check-return-value

File: packages/common/BUILD.bazel (hunk 6 above)
Comment (alexeagle) [reply packages/common/BUILD.bazel]: :+1:

File: packages/common/BUILD.bazel (hunk 15 above)
Comment (alexeagle) [reply packages/common/BUILD.bazel]: yes, they should, assuming there are any decorators that require summary files. I got some errors doing so, but we can push more on them.

Note that this PR is incomplete - there are probably more packages missing build files. The motivation for this PR is the minimal set of changes needed to compile the `packages/core/test/**/*.ts`

File: packages/common/testing/BUILD.bazel (hunk 8 above)
Comment (alexeagle) [reply packages/common/testing/BUILD.bazel]: yes!

File: packages/compiler/test/BUILD.bazel (hunk 10 above)
Comment (alexeagle) [reply packages/compiler/test/BUILD.bazel]: Iterated with `ibazel` until the build was green.

We really need tazel, which my intern  is scheduled to write.

These lists are exhaustive to satisfy strict deps. Note that we now have the option to opt-in or opt-out of strict deps. We should discuss.

File: packages/core/testing/BUILD.bazel (hunk 13 above)
Comment (alexeagle) [reply packages/core/testing/BUILD.bazel]: Our random grab-bag of typings in the packages folder, `goog.d.ts` and `es6-subset.d.ts` and so on

File: packages/platform-browser-dynamic/BUILD.bazel (hunk 16 above)
Comment (alexeagle) [reply packages/platform-browser-dynamic/BUILD.bazel]: We should not need package-specific tsconfigs anymore. The google model, and the recommendation to all teams, is to have one tsconfig for a monorepo

File: packages/platform-server/testing/BUILD.bazel (hunk 19 above)
Comment (alexeagle) [reply packages/platform-server/testing/BUILD.bazel]: really! if you want this to compile, that is...

File: packages/service-worker/BUILD.bazel (hunk 21 above)
Comment (alexeagle) [reply packages/service-worker/BUILD.bazel]: no problem, will catch that in CI

File: packages/tsconfig-build.json (hunk 23 above)
Comment (alexeagle) [reply packages/tsconfig-build.json]: I think we can turn it on now, I just merged a fix for strictdeps lookup paths

File: package.json (hunk 1 above)
Comment (alexeagle) [reply package.json]: Added https://github.com/ReactiveX/rxjs/pull/3131 to commit msg

File: yarn.lock (hunk 24 above)
Comment (alexeagle) [reply yarn.lock]: it's a workaround for an issue I spent all day yesterday landing, see https://github.com/bazelbuild/rules_typescript/pull/95

Comment (alexeagle) [reply yarn.lock]: Workaround is still needed to land this, but I'll keep working on the underlying issue now that I understand it better.

File: packages/common/BUILD.bazel (hunk 15 above)
Comment (IgorMinar) [reply packages/common/BUILD.bazel]: ah. that was not obvious to me.

File: packages/animations/browser/testing/BUILD.bazel (hunk 4 above)
Comment (IgorMinar) [reply packages/animations/browser/testing/BUILD.bazel]: I see. that makes sense. thanks for explaining!

File: packages/platform-browser-dynamic/BUILD.bazel (hunk 16 above)
Comment (IgorMinar) [reply packages/platform-browser-dynamic/BUILD.bazel]: interesting. ok

**Review by IgorMinar — APPROVED** [review]

**Review by chuckjaz — APPROVED** [review]
